docs/design/kata-nydus-design.md

```diff
@@ -0,0 +1,91 @@
+
+The following benchmarking result shows the performance improvement compared with the OCI image for the container cold startup elapsed time on containerd. As the OCI image size increases, the container startup time of using Nydus image remains very short.
+
+![nydus-profile](arch-images/perf.jpeg)
+
+## Proposal - Bring lazyload ability to Kata Containers
+
+Nydusd is a fuse/virtiofs daemon which is provided by nydus project and it supports Passthroughfs and [RAFS](https://github.com/dragonflyoss/image-service/blob/master/docs/nydus-design.md) (Registry Acceleration File System) natively, so in Kata Containers, we can use nydusd in place of virtiofsd and mount nydus image to guest in the meanwhile.
+
+The process of creating/starting Kata Containers with virtiofsd,
+
+1. when creating sandbox, the Kata Containers shim will launch virtiofsd before VM starts, so VM can share directories with host.
+2. when creating container, the Kata Containers shim will mount rootfs to /path/to/bundle/rootfs and bind mount rootfs to `sharedir`, so it can be seen in the guest.
+
+The process of creating/starting Kata Containers with nydusd,
+
+![kata-nydus](arch-images/kata-nydus.png)
+
+1. when creating sandbox, the Kata Containers Containerd v2 shim starts `nydusd` daemon and mounts `passthroughfs` filesystem before the VM starts.
+
+```bash
+# start nydusd
+$ sudo /usr/local/bin/nydusd --log-level info  
+    --sock /run/vc/vm/${sandboxID}/vhost-user-fs.sock 
+    --apisock  /run/vc/vm/${sandboxID}/api.sock
+```
+
+```bash
+# mount passthroughfs
+$ curl -v --unix-socket /run/vc/vm/${sandboxID}/api.sock \
+    -X POST "http://localhost/api/v1/mount?mountpoint=/foo1" -H 'accept: */*' \
+    -H "Content-Type: application/json" \
+    -d '{
+            "source":"/path/to/sharedir",
+            "fs_type":"passthrough_fs",
+            "config":""
+    }'
+```
+
+2. when creating normal container, the Kata Containers Containerd v2 shim send request to nydusd to mount `rafs` to guest. The Kata Containers Containerd v2 shim will also bind mount `snapshotdir` which `nydus-snapshotter` assigns to `sharedir`, so in guest, container rootfs=overlay(lowerdir=`rafs`, upperdir=`snapshotdir/fs`, workdir=`snapshotdir/work`)
+
+```bash
+# mount rafs
+# source: the metafile of nydus image
+# config: the config of this image
+$ curl --unix-socket /run/vc/vm/${sandboxID}/api.sock \
+    -X POST "http://localhost/api/v1/mount?mountpoint=/bar1" -H 'accept: */*' \
+    -H "Content-Type: application/json" \
+    -d '{
+            "source":"/path/to/bootstrap",
+            "fs_type":"rafs",
+            "config":"{}",
```

Comment:
Oh - this `config` is different to the one above as it specifies an empty JSON doc. Can you make them consistent?

Reply:
when type is `passthrough_fs`, `config` is empty, when type is `rafs`, config is a json doc, I have update